internal/dataplane/kongstate/kongstate.go

```diff
@@ -690,3 +683,301 @@ func maybeLogKongIngressDeprecationError(logger logr.Logger, services []*corev1.
 		}
 	}
 }
+
+// FillCustomEntities fills custom entities in KongState.
+func (ks *KongState) FillCustomEntities(
+	logger logr.Logger,
+	s store.Storer,
+	failuresCollector *failures.ResourceFailuresCollector,
+	schemaGetter SchemaGetter,
+	workspace string,
+) {
+	logger = logger.WithName("fillCustomEntities")
+	entities := s.ListKongCustomEntities()
+	if len(entities) == 0 {
+		return
+	}
+
+	if ks.CustomEntities == nil {
+		ks.CustomEntities = map[string]*KongCustomEntityCollection{}
+	}
+	// Fetch relations between plugins and services/routes/consumers and store the pointer to translated Kong entities.
+	// Used for fetching entity referred by a custom entity and fill the ID of referred entity.
+	pluginRels := ks.getPluginRelatedEntitiesRef(s, logger)
+
+	for _, entity := range entities {
+		// reject the custom entity if its type is in "known" entity types that are already processed.
+		if _, isKnownType := KnownEntityTypeMap[entity.Spec.EntityType]; isKnownType {
+			failuresCollector.PushResourceFailure(
+				fmt.Sprintf("cannot use known entity type %s in custom entity", entity.Spec.EntityType),
+				entity,
+			)
+			continue
+		}
+		// Fetch the entity schema.
+		schema, err := ks.fetchEntitySchema(schemaGetter, entity.Spec.EntityType)
+		if err != nil {
+			failuresCollector.PushResourceFailure(
+				fmt.Sprintf("failed to fetch entity schema for entity type %s: %v", entity.Spec.EntityType, err),
+				entity,
+			)
+			continue
+		}
+		// Unmarshal fields of the entity.
+		var parsedEntity map[string]interface{}
+		err = json.Unmarshal(entity.Spec.Fields.Raw, &parsedEntity)
+		if err != nil {
+			failuresCollector.PushResourceFailure(fmt.Sprintf("failed to unmarshal fields of entity: %v", err), entity)
+			continue
+		}
+		// Fill the "foreign" fields if the entity has such fields referencing services/routes/consumers.
+		ks.fillCustomEntityForeignFields(logger, entity, schema, parsedEntity, pluginRels, workspace)
+		// put the entity into custom collection.
+		if _, ok := ks.CustomEntities[entity.Spec.EntityType]; !ok {
+			ks.CustomEntities[entity.Spec.EntityType] = &KongCustomEntityCollection{
+				Schema: schema,
+			}
+		}
+		collection := ks.CustomEntities[entity.Spec.EntityType]
+		collection.Entities = append(collection.Entities, CustomEntity{
+			Object:              parsedEntity,
+			k8sKongCustomEntity: entity,
+		})
+	}
+
+	ks.sortCustomEntities()
+}
+
+// fetchEntitySchema fetches schema of an entity by its type and stores the schema in its custom entity collection
+// as a cache to avoid excessive calling of Kong admin APIs.
+func (ks *KongState) fetchEntitySchema(schemaGetter SchemaGetter, entityType string) (EntitySchema, error) {
+	collection, ok := ks.CustomEntities[entityType]
+	if ok {
+		return collection.Schema, nil
+	}
+	schema, err := schemaGetter.Get(context.Background(), entityType)
```

Comment:
I wanted to suggest context plumbing to be done but I'm surprised to see `BuildKongConfig` not having `context.Context` as parameter. In that case I'd say `context.Background()` is OK here.